src/rules/mapreduce.jl

```diff
@@ -0,0 +1,30 @@
+#####
+##### `map`
+#####
+
+function rrule(::typeof(map), f, xs...)
+    T = Tuple{Core.Typeof(f), map(eltype, xs)...}
+    if hasmethod(rrule, T)
+        body = i->(ȳi, xis...)->begin
+            _, ∂x = rrule(f, xis...)
+            extern((∂x isa Tuple ? ∂x[i] : ∂x)(ȳi))
```

Comment:
#57 will allow us to dump the `isa Tuple` check and just use `∂x[i]` unconditionally

Reply:
Rebased to include that change.